### DISCUSSION TITLE
Recommended patterns for generators?

### DISCUSSION THREAD

**Discussion by trusktr** [eclipse-langium/langium#1078] [Q&A]
Are there any recommended or existing patterns to use for code generators? Namely, any classes to extend from or APIs to use? Or is this area you're-on-your-own?

The code that I'm wondering about is, for example, the code that a Langium user would continue writing to expand the CLI generator (generate command) bootstrapped from the hello-world template.

I believe this is probably where things get the most custom, depending on the user's language, but I want to know if there is at least a list of things recommended to keep in mind so that we are on Langium's best practices path.

So far, this is what ByteScript's generator is looking like (very naive at the moment):

https://github.com/bytescript/bytescript/blob/wasm-output-foundation-construction/src/cli/generator.ts

([Permalink at time of writing](https://github.com/bytescript/bytescript/blob/2c60c53ce1b32f0f2af74e80483ed35c5b63fecf/src/cli/generator.ts#LL1C2-L1C2))

The command `npm install && npm test` while on this (temporary) `wasm-output-foundation-construction` branch will compile the `poc1.ts` file to a `poc1.wasm` WebAssembly file, and run it in Node.js and Chrome headless.

**Comment by msujew** [comment]
> Are there any recommended or existing patterns to use for generators? Namely, any classes to extend from or APIs to use? Or is this area you're-on-your-own?

We do have an API for generating text, see [here](https://www.typefox.io/blog/code-generation-for-langium-based-dsls). It's optimized for generating formatted code. Otherwise, most things related to code generation depend so much on the actual semantics of the language, that there's very little in common between generators.

Most generators in our experience are very similar to tree-walking interpreters. I.e. you walk through your AST, and transform it into another representation. This works well for source to target transformations where the paradigm doesn't change, but can become more complex in case it does.

**Comment by spoenemann** [comment]
We recently published a blog post about this, and two more posts are going to follow soon:
https://www.typefox.io/blog/code-generation-for-langium-based-dsls

**Reply by trusktr** [reply]
@msujew @spoenemann thank you both. I also saw and tried the `CompositeGeneratorNode`, as an alternative to constructing strings. I have in fact also made my own indentation-fixing template tag functions before. Was curious if there was anything more high level, but as mentioned, it can be very specific per language so perhaps not.

**Reply by spoenemann** [reply]
Part two published today: https://www.typefox.io/blog/code-generation-for-langium-based-dsls-2

**Reply by trusktr** [reply]
Nice. Really curious about tracing in part 3!

**Reply by sailingKieler** [reply]
@trusktr In case you didn't notice: https://www.typefox.io/blog/code-generation-for-langium-based-dsls-3 is available in the meantime.
